flang/lib/Lower/HostAssociations.cpp

```diff
@@ -7,19 +7,430 @@
+};
+
+/// Struct to be used as argument in walkCaptureCategories when building the
+/// tuple element value for a host associated variable.
+struct PlaceInTuple {
+  /// walkCaptureCategories returns nothing.
+  using Result = void;
+  /// Value of the variable inside the host procedure.
+  fir::ExtendedValue hostValue;
+  /// Address of the tuple element of the variable.
+  mlir::Value addrInTuple;
+  mlir::Location loc;
+};
+
+/// Struct to be used as argument in walkCaptureCategories when instantiating a
+/// host associated variables from its tuple element value.
+struct GetFromTuple {
+  /// walkCaptureCategories returns nothing.
+  using Result = void;
+  /// Symbol map inside the internal procedure.
+  Fortran::lower::SymMap &symMap;
+  /// Value of the tuple element for the host associated variable.
+  mlir::Value valueInTuple;
+  mlir::Location loc;
+};
+
+/// Base class that must be inherited with CRTP by classes defining
+/// how host association is implemented for a type of symbol.
+/// It simply dispatches visit() calls to the implementations according
+/// to the argument type.
+template <typename SymbolCategory>
+class CapturedSymbols {
+public:
+  template <typename T>
+  static void visit(const T &, Fortran::lower::AbstractConverter &,
+                    const Fortran::semantics::Symbol &,
+                    const Fortran::lower::BoxAnalyzer &) {
+    static_assert(!std::is_same_v<T, T> &&
+                  "default visit must not be instantiated");
+  }
+  static mlir::Type visit(const GetTypeInTuple &,
+                          Fortran::lower::AbstractConverter &converter,
+                          const Fortran::semantics::Symbol &sym,
+                          const Fortran::lower::BoxAnalyzer &) {
+    return SymbolCategory::getType(converter, sym);
+  }
+  static void visit(const PlaceInTuple &args,
+                    Fortran::lower::AbstractConverter &converter,
+                    const Fortran::semantics::Symbol &sym,
+                    const Fortran::lower::BoxAnalyzer &) {
+    return SymbolCategory::placeInTuple(args, converter, sym);
+  }
+  static void visit(const GetFromTuple &args,
+                    Fortran::lower::AbstractConverter &converter,
+                    const Fortran::semantics::Symbol &sym,
+                    const Fortran::lower::BoxAnalyzer &ba) {
+    return SymbolCategory::getFromTuple(args, converter, sym, ba);
+  }
+};
+
+/// Class defining simple scalars are captured in internal procedures.
+/// Simple scalars are non character intrinsic scalars. They are captured
+/// as !fir.ptr<T> (.e.g !fir.ptr<i32> for INTEGER(4)).
+class CapturedSimpleScalars : public CapturedSymbols<CapturedSimpleScalars> {
+public:
+  static mlir::Type getType(Fortran::lower::AbstractConverter &converter,
+                            const Fortran::semantics::Symbol &sym) {
+    return fir::PointerType::get(converter.genType(sym));
+  }
+
+  static void placeInTuple(const PlaceInTuple &args,
```

Comment:
`placeInTuple` is the host-side instantiation of the tuple frame that will be passed, right? Maybe call it `instantiateHostTuple`?